feat(connector): add QQ connector

yenharvey opened the pull request:
<!--
  For non-English users:
  It's okay to post in your language, but remember to use English for the body (you can paste the result of Google Translate), and put everything else as attachments.
  Issues with a non-English body will be DIRECTLY CLOSED until it's updated.
-->

<!-- MANDATORY -->
## Summary
<!-- Provide detailed PR description below -->
Add qq connector.

<!-- MANDATORY -->
## Testing
<!-- How did you test this PR? -->
I have manually tested the QQ connector by:
- Setting up the QQ connector in the Logto management console with the necessary App ID and App Secret.
- Initiating the sign-in flow with the QQ connector on a test application.
- Successfully authenticating with a QQ account and being redirected back to the application with user information.
- Tested the sign-up flow with a new QQ account.
- Verified the user information retrieved from QQ (e.g., nickname, avatar).

<!-- MANDATORY -->
## Checklist
<!-- The palest ink is better than the best memory -->

- [x] `.changeset`
- [x] unit tests
- [x] integration tests
- [x] necessary TSDoc comments

![image](https://github.com/user-attachments/assets/df30328e-c7fc-4327-82be-f125a22ebcb0)
![image](https://github.com/user-attachments/assets/41ab77d7-033e-444e-bcb2-ab508a155c0f)
![image](https://github.com/user-attachments/assets/26699d6f-519b-4f68-80f5-494bfb0c7181)

yenharvey committed:
feat: add QQ social connector for Web. Refs #6769

Signed-off-by: Yen Harvey <2117555041@qq.com>

yenharvey committed:
test: fix QQ connector tests

Signed-off-by: Yen Harvey <2117555041@qq.com>

yenharvey committed:
chore: add changeset for QQ connector

Signed-off-by: Yen Harvey <2117555041@qq.com>

cheezone commented:
Hi @YenHarvey @darcyYe, just a gentle follow-up on this PR. The changes here would be really helpful for our current workflow, and we’d greatly appreciate your review when you have a moment. Thanks so much for your time and expertise!

cheezone commented:
@YenHarvey Noticed the QQ logo image is quite large and takes ~3s to load on 4G. Could we optimize it (e.g., compress/resize) for faster mobile loading? Thanks!

yenharvey committed:
feat: optimize QQ connector logo SVG size

YenHarvey commented:
@cheezone I've reduced the logo size to 2.5k.

darcyYe commented on packages/connectors/connector-qq/CHANGELOG.md:
no need for this file, duplicated with changeset file. The changelog will be generated automatically.

darcyYe commented on packages/connectors/connector-qq/README.en.md:
it could be better if we merge the content of these two README.md file. And we prefer to keep EN README.md content before Chinese guides. You can refer to other connectors such as DingTalk.

charIeszhao replied:
Yes, EN version should be default.

darcyYe commented on packages/connectors/connector-qq/package.json:
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "@logto/connector-qq",
+  "version": "1.0.0",
```
use 0.1.0 or 0.0.0 as version number since the version number will automatically bump along with changeset file.

darcyYe commented on packages/connectors/connector-qq/src/constant.ts:
```diff
@@ -0,0 +1,52 @@
+import type { ConnectorMetadata } from '@logto/connector-kit';
+import { ConnectorPlatform, ConnectorConfigFormItemType } from '@logto/connector-kit';
+
+export const authorizationEndpoint = 'https://graph.qq.com/oauth2.0/authorize';
+export const accessTokenEndpoint = 'https://graph.qq.com/oauth2.0/token';
+export const userInfoEndpoint = 'https://graph.qq.com/user/get_user_info';
+export const openIdEndpoint = 'https://graph.qq.com/oauth2.0/me';
+
+export const defaultScope = 'get_user_info,get_unionId';
```
it is better to leave comments showing these endpoint information so that maintainer and review can easily find corresponding reference.

darcyYe commented on packages/connectors/connector-qq/src/index.ts:
Also, please add some comments and docs reference to those API calls.

darcyYe commented:
Thanks for you contribution

yenharvey committed:
docs: add detailed API comments and optimize documentation

YenHarvey commented:
I've addressed all the feedback by adding detailed API comments to the endpoints and functions and removed the duplicate CHANGELOG.md file as it will be auto-generated by the changeset. Thank you!

charIeszhao commented on packages/connectors/connector-qq/README.zh.md:
Could you please rename it to `README.zh-CN.md`? We have similar README structure in connectors such as `connector-xiaomi` and `connector-yunpian-sms`.

charIeszhao commented on packages/connectors/connector-qq/README.md:
```diff
@@ -0,0 +1,50 @@
+# QQ Social Connector
+
+The official Logto connector for QQ social sign-in [中文文档](https://github.com/logto-io/logto/tree/master/packages/connectors/connector-qq/README.zh.md)
```
```suggestion
The official Logto connector for QQ social sign-in [中文文档](https://github.com/logto-io/logto/tree/master/packages/connectors/connector-qq/README.zh-CN.md)
```

charIeszhao commented on packages/connectors/connector-qq/README.md:
```diff
@@ -0,0 +1,50 @@
+
+QQ is a social platform by Tencent with over 600 million users. This connector helps end-users sign in to your application using their QQ accounts.
+
+## Get Started
+
+1. Create a developer account on [QQ Connect platform](https://connect.qq.com/)
+2. Access the [Application Management](https://connect.qq.com/manage.html)
+3. Create a new application (if you don't already have one)
+
+## Configure QQ Connect Application
+
+1. Access the [Application Management](https://connect.qq.com/manage.html)
+2. Configure the OAuth settings:
+   - Open the application you want to use for sign-in, and click on "Application Info"
+   - Add "Website Callback Domain": `logto_endpoint`
+   - Add "Website Callback URL" : `${logto_endpoint}/callback/${connector_id}`
+3. Get the `APP ID` and `APP Key` from the application info page
+4. Fill in the values from step 3 to the `clientId` and `clientSecret` fields in the Logto Admin Console
+
+## Permission Requirements
+
+To use this connector, you need to apply for the following permissions on the QQ Connect platform:
+
+1. **UnionID Permission**: You must apply for the UnionID API permission to ensure we can obtain the user's unique identifier.
+   Please apply for this permission through the [QQ Connect platform](https://connect.qq.com/).
+
+2. **Default scope**: `get_user_info`
+   This is used to obtain the user's basic information, such as nickname and avatar.
+   By default, if no scope parameter is set, the system will use this permission.
+
+## Test QQ Connector
+
+That's it! Don't forget to enable the connector in the [Sign-in Experience](https://{logto_endpoint}/console/connectors/social).
+
+## References
```
We prefer using sentense case whenever possible in headings.
```suggestion
## Get started

1. Create a developer account on [QQ Connect platform](https://connect.qq.com/)
2. Access the [Application Management](https://connect.qq.com/manage.html)
3. Create a new application (if you don't already have one)

## Configure QQ Connect application

1. Access the [Application Management](https://connect.qq.com/manage.html)
2. Configure the OAuth settings:
   - Open the application you want to use for sign-in, and click on "Application Info"
   - Add "Website Callback Domain": `<your-logto-endpoint>`
   - Add "Website Callback URL" : `<your-logto-endpoint>/callback/<your-connector-id>`
3. Get the `APP ID` and `APP Key` from the application info page
4. Fill in the values from step 3 to the `clientId` and `clientSecret` fields in the Logto Admin Console

## Permission requirements

To use this connector, you need to apply for the following permissions on the QQ Connect platform:

1. **UnionID Permission**: You must apply for the UnionID API permission to ensure we can obtain the user's unique identifier.
   Please apply for this permission through the [QQ Connect platform](https://connect.qq.com/).

2. **Default scope**: `get_user_info`
   This is used to obtain the user's basic information, such as nickname and avatar.
   By default, if no scope parameter is set, the system will use this permission.

## Test QQ connector

That's it! Don't forget to enable the connector in the [Sign-in Experience](https://{logto_endpoint}/console/connectors/social).

## References
```

charIeszhao commented on packages/connectors/connector-qq/src/mock.ts:
```diff
@@ -0,0 +1,40 @@
+  ret: 0,
+  msg: '',
+  is_lost: 0,
+  nickname: 'Ciallo',
+  gender: '男',
+  gender_type: 2,
+  province: '北京',
+  city: '海淀',
+  year: '1995',
+  figureurl: 'https://thirdqq.qlogo.cn/ek_qqapp/user-small.jpg',
+  figureurl_1: 'https://thirdqq.qlogo.cn/ek_qqapp/user-small.jpg',
+  figureurl_2: 'https://thirdqq.qlogo.cn/ek_qqapp/user-large.jpg',
+  figureurl_qq: 'https://thirdqq.qlogo.cn/ek_qqapp/user-qq-full.jpg',
+  figureurl_qq_1: 'https://thirdqq.qlogo.cn/ek_qqapp/user-small.jpg',
+  figureurl_qq_2: 'https://thirdqq.qlogo.cn/ek_qqapp/user-large.jpg',
```
It would be better to use dummy data instead of meaningful informations in the mock response.

charIeszhao commented on packages/connectors/connector-qq/src/index.ts:
```diff
@@ -0,0 +1,271 @@
+        throw new ConnectorError(ConnectorErrorCodes.InvalidResponse, result.error);
+      }
+
+      // Ok?
```
Please either update or remove the umbiguous comment.

charIeszhao commented on packages/connectors/connector-qq/package.json:
```diff
@@ -0,0 +1,69 @@
+    "lib",
+    "docs",
+    "logo.svg",
+    "logo-dark.svg"
```
No `logo-dark.svg` provided. So we can just remove it from the files list for now.

yenharvey committed:
docs: update heading format to sentence case and refine documentation

charIeszhao reviewed (commented):
I reviewed the updated changes and it all looks good to me now. Thanks again for the contribution and quick response.

@darcyYe Leaving it to you for the final approval.

charIeszhao commented:
There seems to be a frozen lockfile issue in your branch. Please take a look at the CI error logs and execute the suggested CLI commands. @YenHarvey

yenharvey committed:
chore: update package.json and pnpm lockfile for consistency

YenHarvey commented:
@charIeszhao I've executed the suggested command from the CI logs: pnpm install --no-frozen-lockfile to update the lockfile. I noticed that 'logo-dark.svg' was automatically added back during this process. To avoid any further issues, I've decided to keep it rather than removing it again. The updated lockfile and package.json have been pushed to the branch.

YenHarvey commented:
@cheezone Yes, this is my first time submitting a PR. I initially started development on the main branch, but later learned it's better practice to use a feature branch for PR development. So I created this new PR with all the changes properly implemented on a dedicated branch. All review comments have been addressed in this PR.  I learn the proper workflow. The PR should be ready for review now. Thanks!

cheezone commented:
@darcyYe Hello, could you please ask if there's a chance for this PR to be reviewed in the near future? This change is quite helpful for me as well, thank you for your hard work!

darcyYe reviewed (approved):
LGTM

gao-sun reviewed (approved):
lgtm on changeset

YenHarvey commented:
Can you provide the logs? It was working fine when I tested it locally and on the server.




SummerNight
***@***.***

YenHarvey commented:
Is your QQ Connect application review already approved? Also, after approval, you need to manually apply for UnionID permission—have you done this step?




SummerNight
***@***.***

cheezone commented:
@YenHarvey 

It's okay, it's resolved. It was just about the UnionID. However, the QQ connector seems strange to me. It might be a problem with QQ. I did a test on the mobile device. It can't launch the QQ mobile app. Instead, it shows a desktop version page asking the user to scan a QR code. After the user scans the code and agrees, it fails to redirect successfully.

YenHarvey commented:
@cheezone Scared me! I thought we were having that awkward "Works only on my machine" situation. This behavior on mobile is actually normal, as my implementation is targeted for web development. I remember mobile needs an extra parameter passed in. Since I wasn't sure if logto had an API to detect platforms, I just temporarily solved the issue of having a QQ connector😜.
